internal/collectors/builder.go

```diff
@@ -0,0 +1,448 @@
+/*
+Copyright 2018 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collectors
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	apps "k8s.io/api/apps/v1beta1"
+	autoscaling "k8s.io/api/autoscaling/v2beta1"
+	batchv1 "k8s.io/api/batch/v1"
+	batchv1beta1 "k8s.io/api/batch/v1beta1"
+	"k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	policy "k8s.io/api/policy/v1beta1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	coll "k8s.io/kube-state-metrics/pkg/collector"
+	metricsstore "k8s.io/kube-state-metrics/pkg/metrics_store"
+	"k8s.io/kube-state-metrics/pkg/options"
+	"k8s.io/kube-state-metrics/pkg/whiteblacklist"
+)
+
+// Builder helps to build collectors. It follows the builder pattern
+// (https://en.wikipedia.org/wiki/Builder_pattern).
+type Builder struct {
```

Comment:
Can we make the Builder a public builder?
1. Instead of have a builder func for every resource, we can have a generic func like this?

```
func (b *Builder) BuildCollector(metrics []metrics.FamilyGenerator,expectedType interface{},
	listWatchFunc func(apiserver string, kubeconfig string, ns string) cache.ListWatch) *collectors.Collector{

	filteredMetricFamilies := filterMetricFamilies(b.whiteBlackList, metrics)
	composedMetricGenFuncs := composeMetricGenFuncs(filteredMetricFamilies)

	familyHeaders := extractMetricFamilyHeaders(filteredMetricFamilies)

	store := metricsstore.NewMetricsStore(
		familyHeaders,
		composedMetricGenFuncs,
	)
	reflectorPerNamespace(b.ctx, expectedType, store,
		b.apiserver, b.kubeconfig, b.namespaces, listWatchFunc)

	return collectors.NewCollector(store)
}
```
2. and add a func 'WithAvailableCollector` instead of an global private var `availableCollectors` 
3. remove WithKubeClient func, instead add func WithApiserver, WithKubeconfig, as openshift have different client for different api group. like this https://github.com/wanghaoran1988/openshift-state-metrics/blob/ce6b3351e63e92f3eb8aa38edd926a65949baecb/pkg/collectors/builder.go#L63

Reply:
Following the [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) I feel like each component using kube-state-metrics as a library should write their own. Its core task is to tie all components together, hence being very specific to the end result. I feel like a generic builder is to abstract. Instead I would prefer a _little copying_ hence reducing abstractions.

What are your thoughts?

Reply:
That's fair